### PR TITLE
Add lock renumbering

### PR DESCRIPTION
**Pull request by mheon**
Still needs Podman frontend, plus a bit more lock code.

Needed for a 1.1 release. We also need to modify RPM postinstall to call this for any 1.0->1.1 upgrades to ensure all containers get migrated properly to SHM locks.

Still messy but feel free to review while I hack the last bits into place.

### CODE REVIEW

**Comment by mheon** [comment]
(Hopefully the main body of the code is pretty clean - this is a rewrite of an earlier effort that turned into a sprawling mess)

**Comment by mheon** [comment]
Alright, code should all be here. Still need to do some testing (and see if I can add tests without a massive hassle), but I'll remove the WIP tag, this is ready for review.

**Comment by mheon** [comment]
Squashed some of the tangential commits together so build_each_commit won't time out

**Comment by mheon** [comment]
Works in all my local testing. It even seems to be same to run while other Podman instances are going, which I wasn't expecting. I really wouldn't recommend doing that, but if you do nothing seems to explode.

@rhatdan @baude @TomSweeneyRedHat @vrothberg PTAL

**Comment by mheon** [comment]
(Still needs manpages + completions. I don't think we can test this in CI right now - it requires a 1.0 or lower Podman to actually do anything)

**Comment by mheon** [comment]
To clarify, you need both a 1.0 or lower Podman, and the latest master with renumber patches, to actually test

**Comment by rhatdan** [comment]
Why change 100 interfaces to all call false, and only one to call true?  Wouldn't it have been easier to add one new interface?

**Comment by mheon** [comment]
This is what happens when I code before coffee. Fixed.

**Comment by mheon** [comment]
bot, retest this please

File: libpod/lock/in_memory_locks.go [hunk 1]
```diff
@@ -89,3 +89,12 @@ func (m *InMemoryManager) RetrieveLock(id uint32) (Locker, error) {
...
 }
+
+// FreeAllLocks frees all locks
+func (m *InMemoryManager) FreeAllLocks() error {
+	for _, lock := range m.locks {
+		lock.allocated = false
+	}
+
+	return nil
+}
```
Comment (TomSweeneyRedHat) [inline libpod/lock/in_memory_locks.go]: This just feels so incredibly dangerous...  Is there anyway to wait for all locks to be freed and while waiting ensure others aren't taken out?

Comment (mheon) [reply libpod/lock/in_memory_locks.go]: We're holding the alive lock at this point, so no other processes are starting or stopping. We also have the SHM lock, so nobody else is allocating/removing at that moment.

Long-lived podman processes will retain the old lock, which means they won't break, but also probably have a different lock than newer processes. In my testing, that doesn't seem to cause problems, but I certainly don't recommend it.

In short, I think we need to mention on the man page to try and avoid using this at the same time as other Podman processes are alive, but we have protection against this doing anything nasty - the worst case is that we break any Podman processes that were running before renumber does, while launching new Podman commands works as expected.

File: libpod/state.go [hunk 2]
```diff
@@ -97,6 +97,30 @@ type State interface {
 	// returned.
 	AllContainers() ([]*Container, error)
 
+	// PLEASE READ FULL DESCRIPTION BEFORE USING.
+	// Rewrite a container's configuration.
+	// This function breaks libpod's normal prohibition on a read-only
+	// configuration, and as such should be used EXTREMELY SPARINGLY and
+	// only in very specific circumstances.
+	// Specifically, it is ONLY safe to use thing function to make changes
```
Comment (TomSweeneyRedHat) [inline libpod/state.go]: s/thing/this/

File: libpod/lock/in_memory_locks.go (hunk 1 above)
Comment (TomSweeneyRedHat) [reply libpod/lock/in_memory_locks.go]: OK, that makes me feel better, but there are no checks in this function that validates that the alive lock and SHM lock are held.  Is there anyway we can check that fact as a safety precaution?  I'm more concerned about future development and someone not getting the locks correctly before calling this.

Comment (mheon) [reply libpod/lock/in_memory_locks.go]: It probably needs the same extended disclaimer I gave to the configuration writer functions - don't ever use this unless you have an extremely good reason and have taken the appropriate precautions

File: cmd/podman/system_renumber.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+var (
+	renumberCommand     cliconfig.SystemRenumberValues
+	renumberDescription = `
+        podman system renumber
+
+        Migrate lock numbers to handle a change in maximum number of locks
```
Comment (vrothberg) [inline cmd/podman/system_renumber.go]: I miss some explanation of why and when this is necessary. Is it something users must to do or an optional migration? The more details the better :)

File: libpod/lock/lock.go [hunk 4]
```diff
@@ -24,6 +24,9 @@ type Manager interface {
 	// The underlying lock MUST be the same as another other lock with the
 	// same UUID.
 	RetrieveLock(id uint32) (Locker, error)
+	// FreeAllLocks frees all allocated locks, in preparation for lock
+	// reallocation.
+	FreeAllLocks() error
```
Comment (vrothberg) [inline libpod/lock/lock.go]: The disclaimer may be worth adding here as well with some notes why and when it's necessary to be called (mostly thinking about godoc.org).

File: libpod/runtime.go [hunk 5]
```diff
@@ -88,12 +90,15 @@ type Runtime struct {
...
-	lock            sync.RWMutex
 	imageRuntime    *image.Runtime
 	firewallBackend firewall.FirewallBackend
 	lockManager     lock.Manager
 	configuredFrom  *runtimeConfiguredFrom
+
+	doRenumber bool
+
+	valid bool
+	lock  sync.RWMutex
```
Comment (vrothberg) [inline libpod/runtime.go]: Can we add comments to the two booleans? The other fields are more or less self-explanatory (as the types can be looked up).

File: libpod/runtime_renumber.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+)
+
+// renumberLocks reassigns lock numbers for all containers and pods in the
+// state.
+// TODO: It would be desirable to make it impossible to call this until all
+// other libpod sessions are dead.
+// Possibly use a read-write file lock, with all non-renumber podmans owning the
+// lock as read, renumber attempting to take a write lock?
+// The alternative is some sort of session tracking, and I don't know how
+// reliable that can be.
```
Comment (vrothberg) [inline libpod/runtime_renumber.go]: We could use the reader-writer locks from containers/storage for that but had to move them to a dedicated package.

File: libpod/boltdb_state.go [hunk 7]
```diff
@@ -783,6 +783,50 @@ func (s *BoltState) AllContainers() ([]*Container, error) {
 	return ctrs, nil
 }
 
+// RewriteContainerConfig rewrites a container's configuration.
+// WARNING: This function is DANGEROUS. Do not use without reading the full
+// comment on this function in state.go.
+func (s *BoltState) RewriteContainerConfig(ctr *Container, newCfg *ContainerConfig) error {
+	if !s.valid {
+		return ErrDBClosed
```
Comment (vrothberg) [inline libpod/boltdb_state.go]: I wonder why `ErrDBClosed` is used. Is the deduction invalid -> database is closed always true? I mostly wonder because of `defer s.closeDBCon(db)` in line 807.

**Comment by mheon** [comment]
@vrothberg The locks were only used in one place, RemoveVolume (and not even everywhere there - there were cases where we'd remove without locks), so rather than try and add them to renumber I just removed them.. They didn't seem to be doing anything

File: libpod/runtime_renumber.go (hunk 6 above)
Comment (mheon) [reply libpod/runtime_renumber.go]: I'm thinking that's probably followon work... I like the idea of locks, but I also really hate that we'd block until every one was released, with no way of telling what was going on. Perfect world, we'd have a way of inspecting the lock and telling that "PIDs 123, 456, 789 are all running, please close before running renumber"

File: libpod/boltdb_state.go (hunk 7 above)
Comment (mheon) [reply libpod/boltdb_state.go]: Valid is... less meaningful than it used to be, in bolt? We initially maintained a single connection, which we reused, and then valid made sense (if valid == false, the connection was closed). Now, we make a new connection for each function, which means valid can be true, and no connection is open... In the current implementation, valid == false mostly means we've "shut down" the database, and are refusing to make new connections

Comment (vrothberg) [reply libpod/boltdb_state.go]: Thanks for insights!

**Comment by vrothberg** [comment]
Thanks for adding the comments, @mheon. Will a manpage be added with a later commit?

**Comment by mheon** [comment]
I was meaning to do that yesterday, but it completely slipped my mind and I just hit review comments. Oops. Manpage incoming.

**Comment by mheon** [comment]
Manpages finally added, managed to get distracted again.

@rhatdan @vrothberg @baude PTAL

**Comment by mheon** [comment]
Squashed more so build_each_commit doesn't time out.

@rhatdan @baude @vrothberg @TomSweeneyRedHat @giuseppe PTAL, let's get this merged for 1.1.0

**Comment by vrothberg** [comment]
@mheon, sorry for not noticing this earlier. libpod.conf doesn't have an option for this now right? It's manpage is also missing an entry.

Besides that LGTM.

**Comment by mheon** [comment]
@vrothberg https://github.com/containers/libpod/pull/2362

File: cmd/podman/libpodruntime/runtime.go [hunk 8]
```diff
@@ -8,8 +8,17 @@ import (
 	"github.com/pkg/errors"
 )
 
+// GerRuntimeRenumber gets a libpod runtime that will perform a lock renumber
```
Comment (TomSweeneyRedHat) [inline cmd/podman/libpodruntime/runtime.go]: I'm surprised validate didn't choke on this.  s/GerRuntimeRenumber/GetRuntimeRenumber/

Comment (mheon) [reply cmd/podman/libpodruntime/runtime.go]: Fixed
